### PR TITLE
feat(a11y/e2e) : axe-selenium for automated tests

### PR DESCRIPTION
**Pull request by jsomsanith-tlnd**
**What is the problem this PR is trying to solve?**
No tools to automate tests on a11y.

**What is the chosen solution to this problem?**
Selenium library to help to run a11y checks easily

See [README](https://github.com/Talend/ui/blob/ba90fa86c0285a3fd1454fe01042e143ebbde0b3/packages/e2e/axe-selenium/README.md)

On error, we save the axe report (json) and a formatted violation report (text).
Next step is to make a simple webapp that consumes the json axe report and display it in a user friendly way.

**Please check if the PR fulfills these requirements**

* [x] The PR commit message follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md)
* [x] Tests for the changes have been added (for bug fixes / features) And [non reg](./screenshots.md) done before need review
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] Related design / discussions / pages (not in jira), if any, are all linked or available in the PR

**[ ] This PR introduces a breaking change**

### CODE REVIEW

File: packages/e2e/axe-selenium/src/main/java/org/talend/axeselenium/Reporter.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+                    all.put(none.getJSONObject(k));
+                }
+
+                appendFixes(sb, all, "Fix all of the following:");
+                appendFixes(sb, node.getJSONArray("any"), "Fix any of the following:");
+            }
+        }
+
+        return sb.toString();
+    }
```
Comment (talend-glorieux) [inline packages/e2e/axe-selenium/src/main/java/org/talend/axeselenium/Reporter.java]: Isn't it possible to use `java.text.MessageFormat` or another template library to properly build that report?
Understanding the end result by looking at each line is impossible.

File: packages/e2e/axe-selenium/src/main/java/org/talend/axeselenium/Reporter.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+     * Writes a raw object out to a JSON file with the specified name.
+     *
+     * @param name   Desired filename, sans extension
+     * @param output Object to write. Most useful if you pass in either the Builder.analyze() response or the
+     *               violations array it contains.
+     */
+    public void writeResults(final String name, final Object output) {
+        try (PrintWriter out = new PrintWriter(name + ".json")) {
+            out.println(output.toString());
+        } catch (final IOException ignored) {}
```
Comment (talend-glorieux) [inline packages/e2e/axe-selenium/src/main/java/org/talend/axeselenium/Reporter.java]: Handle the exception.

File: packages/e2e/axe-selenium/src/main/java/org/talend/axeselenium/Reporter.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    /**
+     * Writes a string report out to a text file with the specified name.
+     *
+     * @param name   Desired filename, sans extension
+     * @param output Report to write
+     */
+    public void writeResults(final String name, final String output) {
+        try (PrintWriter out = new PrintWriter(name + ".txt")) {
+            out.println(output);
+        } catch (final IOException ignored) {}
```
Comment (talend-glorieux) [inline packages/e2e/axe-selenium/src/main/java/org/talend/axeselenium/Reporter.java]: Handle the exception.

File: packages/e2e/axe-selenium/src/main/java/org/talend/axeselenium/Script.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            while ((line = reader.readLine()) != null) {
+                sb.append(line);
+                sb.append(lineSeparator);
+            }
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        } finally {
+            if (reader != null) {
+                try { reader.close(); }
+                catch (IOException ignored) {}
```
Comment (talend-glorieux) [inline packages/e2e/axe-selenium/src/main/java/org/talend/axeselenium/Script.java]: Handle the exception.

File: packages/e2e/axe-selenium/src/main/java/org/talend/axeselenium/axe/Axe.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    public String reportViolations(final String reportFolder, final String name, final JSONObject result) {
+        new File(reportFolder).mkdirs();
+        final JSONArray violations = result.getJSONArray("violations");
+        if (violations.length() > 0) {
+            final String formattedViolations = reporter.report(violations);
+            reporter.writeResults(reportFolder + '/' + name, result);
+            reporter.writeResults(reportFolder + '/' + name, formattedViolations);
+            return formattedViolations;
+        }
+        return null;
```
Comment (talend-glorieux) [inline packages/e2e/axe-selenium/src/main/java/org/talend/axeselenium/axe/Axe.java]: I'd rather see `""` being returned to keep the returned type consistent.

File: packages/e2e/axe-selenium/src/test/resources/test-server.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+http.createServer(function (req, res) {
+    const urlParts = url.parse(req.url, true);
+    const urlParams = urlParts.query;
+    const valid = urlParams && urlParams.valid;
+
+	res.writeHead(200, { 'Content-Type': 'text/html' });
+	res.end(valid === 'true' ? validHTML : invalidHTML);
+
+}).listen('5005');
```
Comment (talend-glorieux) [inline packages/e2e/axe-selenium/src/test/resources/test-server.js]: Missing newline.

File: packages/e2e/axe-selenium/src/main/java/org/talend/axeselenium/axe/Axe.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+     * @param reportFolder The folder where to store the scan result.
+     * @param name The result file name.
+     * @param result The scan result.
+     * @return the formatted violation report string, null if no violation.
+     */
+    public String reportViolations(final String reportFolder, final String name, final JSONObject result) {
+        new File(reportFolder).mkdirs();
+        final JSONArray violations = result.getJSONArray("violations");
+        if (violations.length() > 0) {
+            reporter.writeResults(reportFolder + '/' + name, result);
```
[the remark this answers is not in the document]
Comment (jsomsanith) [reply packages/e2e/axe-selenium/src/main/java/org/talend/axeselenium/axe/Axe.java]: Use `Paths.get()`

File: packages/e2e/axe-selenium/src/main/java/org/talend/axeselenium/Reporter.java (hunk 1 above)
Comment (jsomsanith) [reply packages/e2e/axe-selenium/src/main/java/org/talend/axeselenium/Reporter.java]: Replace it with Formatter, bound to the StringBuilder, but it doesn't improve readability :/

File: packages/e2e/axe-selenium/src/main/java/org/talend/axeselenium/Reporter.java (hunk 2 above)
Comment (jsomsanith) [reply packages/e2e/axe-selenium/src/main/java/org/talend/axeselenium/Reporter.java]: 👍  I let the method throw the exception to let the test fail.

File: packages/e2e/axe-selenium/src/main/java/org/talend/axeselenium/Reporter.java (hunk 3 above)
Comment (jsomsanith) [reply packages/e2e/axe-selenium/src/main/java/org/talend/axeselenium/Reporter.java]: Same here

File: packages/e2e/axe-selenium/src/main/java/org/talend/axeselenium/Script.java (hunk 4 above)
Comment (jsomsanith) [reply packages/e2e/axe-selenium/src/main/java/org/talend/axeselenium/Script.java]: Same here

File: packages/e2e/axe-selenium/src/main/java/org/talend/axeselenium/axe/Axe.java (hunk 5 above)
Comment (jsomsanith) [reply packages/e2e/axe-selenium/src/main/java/org/talend/axeselenium/axe/Axe.java]: It's an indicator that there are no violation. `violations == null` vs `violations.equals("")`.
I let the method return null.

File: packages/e2e/axe-selenium/src/test/resources/test-server.js (hunk 6 above)
Comment (jsomsanith) [reply packages/e2e/axe-selenium/src/test/resources/test-server.js]: Added

**Comment by frassinier** [comment]
🗑 ?

**Comment by jsomsanith-tlnd** [comment]
Let's keep this one, I still have hope :D

**Comment by jsomsanith-tlnd** [comment]
After 2.5 years, QA has still not worked on that.
Let's close

**Pull request closed without merging**